library/ssl_tls12_server.c

```diff
@@ -2589,35 +2589,44 @@ static int ssl_get_ecdh_params_from_cert(mbedtls_ssl_context *ssl)
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
-    unsigned char buf[
-        PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE(PSA_VENDOR_ECC_MAX_CURVE_BITS)];
+    mbedtls_pk_context *pk;
+    mbedtls_pk_type_t pk_type;
     psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+#if !defined(MBEDTLS_PK_USE_PSA_EC_DATA)
     uint16_t tls_id = 0;
     psa_ecc_family_t ecc_family;
     size_t key_len;
-    mbedtls_pk_context *pk;
     mbedtls_ecp_group_id grp_id;
+    unsigned char buf[
+        PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE(PSA_VENDOR_ECC_MAX_CURVE_BITS)];
+    mbedtls_ecp_keypair *key;
+#endif /* !MBEDTLS_PK_USE_PSA_EC_DATA */
 
     pk = mbedtls_ssl_own_key(ssl);
 
     if (pk == NULL) {
         return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
     }
 
-#if !defined(MBEDTLS_PK_USE_PSA_EC_DATA)
-    mbedtls_ecp_keypair *key = mbedtls_pk_ec_rw(*pk);
-#endif /* !MBEDTLS_PK_USE_PSA_EC_DATA */
+    pk_type = mbedtls_pk_get_type(pk);
 
-    switch (mbedtls_pk_get_type(pk)) {
+    switch (pk_type) {
         case MBEDTLS_PK_OPAQUE:
+#if defined(MBEDTLS_PK_USE_PSA_EC_DATA)
+        case MBEDTLS_PK_ECKEY:
+        case MBEDTLS_PK_ECKEY_DH:
+        case MBEDTLS_PK_ECDSA:
+#endif /* MBEDTLS_PK_USE_PSA_EC_DATA */
             if (!mbedtls_pk_can_do(pk, MBEDTLS_PK_ECKEY)) {
                 return MBEDTLS_ERR_SSL_PK_TYPE_MISMATCH;
             }
 
             ssl->handshake->ecdh_psa_privkey = pk->priv_id;
 
-            /* Key should not be destroyed in the TLS library */
-            ssl->handshake->ecdh_psa_privkey_is_external = 1;
+            if (pk_type == MBEDTLS_PK_OPAQUE) {
```

Comment:
I think this is superfluous: if the key is not opaque, it's owned by the PK context, not TLS, so it shouldn't be destroyed by TLS. (In the opaque case, the key is owned neither by the PK context nor by the TLS handshake structure, but by the application.)

Reply:
Yes, correct, indeed the `ecdh_psa_privkey_is_external` flag is used elsewhere in the code in order to tell if the key is to be destroyed by TLS or not. This seems OK to me, but since you're commenting on this I guess I missed something :)
Can you please try to explain a bit more?

Reply:
I guess "this" in my first comment was ambiguous: I meant the `if (pk_type == MBEDTLS_PK_OPAQUE)` is superfluous, and we should set `ssl->handshake->ecdh_psa_privkey_is_external = 1;` unconditionally here, since the key is never owned by the TLS handshake structure here: it is only be owned by TLS in the `!USE_PSA_EC_DATA` case below, where we create it by calling `psa_import_key()`).

Reply:
Oh, I see, so it's not a matter of opaque/non-opaque and it can still be owned by "the library", but perhaps by some other module.